WIP: added unified validation scheme

martinwimpff opened the pull request:
Added a unified validation scheme example as discussed in  #378 
This tutorial shares some similarities with #349 
@bruAristimunha: let me know what your think about this first draft

bruAristimunha commented:
Hi @martinwimpff,

Thank you so much for your PR!

Here is the render version: 
https://output.circle-artifacts.com/output/job/a34faa77-933e-41f7-8a6f-4905c1fe2ef4/artifacts/0/dev/auto_examples/plot_unified_validation_scheme.html#sphx-glr-auto-examples-plot-unified-validation-scheme-py

I am still revising it, but it is already pretty cool! I will try to add a figure to make it easier to differentiate the three schemes.

sliwy commented on examples/plot_unified_validation_scheme.py:
```diff
@@ -0,0 +1,344 @@
+)
+
+######################################################################
+# We will make use of the sklearn library to do the Hyperparameter
```
```suggestion
# We will make use of the sklearn library to do the hyperparameter
```

bruAristimunha commented on examples/plot_unified_validation_scheme.py:
```diff
@@ -0,0 +1,344 @@
+
+
+######################################################################
+# We can easily split the dataset using additional info stored in the
+# description attribute, in this case ``session`` column. We select
+# ``session_T`` for training and ``session_E`` for validation.
+# For other datasets you might have to choose another column.
```
```suggestion
# We can easily split the dataset using additional info stored in the 
# description attribute, in this case, the  ``session`` column. We select 
# ``session_T``  for training and  ``session_E`` for validation.
# For other datasets, you might have to choose another column.
```

sliwy commented on examples/plot_unified_validation_scheme.py:
```diff
@@ -0,0 +1,344 @@
+train_val_split = [tuple(train_test_split(X_train.indices_, test_size=0.2, shuffle=False))]
+
+######################################################################
+# Define the ``fit_params`` and the ``paramter_grid`` i.e. list all
```
```suggestion
# Define the ``fit_params`` and the ``parameter_grid`` i.e. list all
```

sliwy commented on examples/plot_unified_validation_scheme.py:
```diff
@@ -0,0 +1,344 @@
+
+######################################################################
+# Define the ``fit_params`` and the ``paramter_grid`` i.e. list all
+# Hyperparamters you want to include in your search.
```
```suggestion
# hyperparameters you want to include in your search.
```

bruAristimunha commented on examples/plot_unified_validation_scheme.py:
```diff
@@ -0,0 +1,344 @@
+
+
+######################################################################
+# This is the easiest training scheme to use as the dataset is only
+# splitted in two distinct sets (``train_set`` and ``test_set``).
+# As this method uses no separate validation split it should only be
+# used for the final evaluation of the (previously!) found
+# Hyperparameter configuration.
```
```suggestion
# This is the easiest training scheme to use as the dataset is only 
# split into two distinct sets (``train_set`` and ``test_set``).
# This scheme uses no separate validation split and should only be
#  used for the final evaluation of the (previously!) found
#  Hyperparameter configuration.
```

bruAristimunha commented on examples/plot_unified_validation_scheme.py:
```diff
@@ -0,0 +1,344 @@
+# Hyperparameter configuration.
+#
+# .. warning::
+#    If you make any use of the ``test_set`` during training
+#    (i.e. by using EarlyStopping) there will be data leakage
+#    which will make the reported generalization capability/decoding
+#    performance of your model less credible.
```
```suggestion
#    If you use the  ``test_set`` during training
#    (i.e. by using EarlyStopping), there will be data leakage
#    which will make your model's reported generalization capability/decoding
#    performance less credible.
```

sliwy commented on examples/plot_unified_validation_scheme.py:
```diff
@@ -0,0 +1,344 @@
+# train-val split.
+#
+
+train_val_split = KFold(n_splits=5, shuffle=False)
```
maybe you can also show how to evaluate model on cross-validation with `cross_val_score` - should be one line somewhere, just to show that it can be also used as performance evaluation method and not only hyperparams search

martinwimpff replied:
Good point - I will add that

sliwy commented on examples/plot_unified_validation_scheme.py:
```diff
@@ -0,0 +1,344 @@
+
+preprocessors = [
+    Preprocessor('pick_types', eeg=True, meg=False, stim=False),  # Keep EEG sensors
+    Preprocessor(scale, factor=1e6, apply_on_array=True),  # Convert from V to uV
```
I would use `np.multiply` here to avoid `DeprecationWarning` in the doc

martinwimpff replied:
Agree

martinwimpff replied:
Do you know how to properly do that? The problem with `np.multiply` is that x2 (the factor) is a positional argument and not a keyword argument. Therefore
`Preprocessor(np.multiply, x2=1e6, apply_on_array=True)`
does not work. My current workaround would look like this:
`Preprocessor(lambda data, factor: np.multiply(data, factor), factor=1e6)`

sliwy replied:
good question! :smile:  I just followed the DeprecationWarning message. I think we should remove the DeprecationWarning and abandon the plan of removing `scale` function because in our flow `np.multiply` can't replace it. At least I don't see an easy way as we use keyword arguments in preprocessing (the lambda you've proposed does exactly the same as current `scale` function). 

We may remove the DeprecationWarning here or in another PR if everyone agrees that it won't work with `np.multiply` @robintibor @hubertjb @bruAristimunha

bruAristimunha replied:
Sure @sliwy. After finishing this tutorial we will do a PR to update the warning message.

bruAristimunha commented on examples/plot_unified_validation_scheme.py:
````diff
@@ -0,0 +1,344 @@
+#
+
+######################################################################
+# Usually when developing a new Deep Learning model/method finding
+# the best (or at least a suitable) Hyperparameter configuration makes
+# up a substantial part of the developement process.
+# As stated above, it is not suitable to use the ```test_set``
+# for this Hyperparamter search. Therefore we need a third split, the
+# so called validation set which is a Subset of the ``train_set``.
+# This second option splits the original ``train_set`` only once
+# (instead of k times as in Option 3) to speed up the tuning process.
+# This method should only be preferred over Option 3 if either the
+# training duration is very long or the Hyperparamter search space is
+# very large.
````
```suggestion
# Usually, when developing a new deep learning model/method finding 
# the best (or at least a suitable) hyperparameter configuration makes 
# up a substantial part of the development process. 
# 
#As stated above, using the ```test_set`` 
# for this hyperparameter search is unsuitable. Therefore we need a third split, 
# the so-called validation set, which is a Subset of the train_set. # This second scheme splits the original train_set only once 
# (instead of k times as in scheme 3) to speed up the tuning process. 
#
# This method should only be preferred over scheme 3 if the 
# training duration is very long, or the Hyperparameter search space # is very large.
```

sliwy commented on examples/plot_unified_validation_scheme.py:
```diff
@@ -0,0 +1,344 @@
+models with Braindecode. We will use the BCIC IV 2a dataset as a showcase example,
+however this scheme holds for any supervised trial-based decoding setting.
+As this tutorial will include additional parts of code like loading and preprocessing,
+defining a model, etc. which are not exclusive to this page (compare `Cropped Decoding
+Tutorial <./plot_bcic_iv_2a_moabb.html>`__), feel free to skip these parts.
```
I would link to trialwise example. This also fixes the link to the file.
```suggestion
defining a model, etc. which are not exclusive to this page (compare `Trialwise Decoding
Tutorial <./plot_bcic_iv_2a_moabb_trial.html>`__), feel free to skip these parts.
```

bruAristimunha commented on examples/plot_unified_validation_scheme.py:
```diff
@@ -0,0 +1,344 @@
+# ----------------------------------------------------
+#
+
+######################################################################
+# As mentioned above, using only one validation split might not be
+# sufficient, as there might be a shift in the data distribution.
+# To compensate this, one can run a k-fold Cross Validation, where
+# every sample of the training set is in the validation set once.
+# By averaging over the k validation scores afterwards, you get a
+# very reliable estimate of how the model would perform on unseen
+# data (test set).
+# The implementation is straight forward as we only have to change one
+# line of code compared to Option 2. The ``KFold`` class will split
+# the ``train_set`` into k folds. We will use 5 folds to get a 80-20
+# train-val split.
+#
```
```suggestion
######################################################################
# As mentioned above, using only one validation split might not be
# sufficient, as there might be a shift in the data distribution.
# To compensate for this, one can run a k-fold Cross Validation, where
# every sample of the training set is in the validation set once.
# After averaging over the k validation scores, you get a
# very reliable estimate of how the model would perform on unseen
# data (test set).
# The implementation is straightforward as we only have to change one
# line of code compared to Option 2. The ``KFold`` class will split
# the ``train_set`` into k folds. We will use 5 folds to get an 80-20
# train-validation split.
#
```

sliwy reviewed (changes requested):
Great job @martinwimpff @bruAristimunha! To me the description is clear and I hope people will read it before doing ML for EEG. My comments are mainly typos and some reformulations.

A visualization of the CV schemes would be highly appreciated.

One more thing, now this tutorial and [`plot_hyperparameter_tuning_with_scikit-learn.py` ](https://github.com/braindecode/braindecode/blob/66ddda17875bb4d5b788228dd7a3b0b6117290b8/examples/plot_hyperparameter_tuning_with_scikit-learn.py) are quite similar. Maybe we can add `hyperparameters search` to the title of this one and just keep this tutorial? What do you think?

bruAristimunha reviewed (commented):
Is very good! I tried to give some small suggestions, I tried not to change the text. Feel free to accept or not the suggestions. I'll try to put some image to illustrate the options.

bruAristimunha commented:
Hi @martinwimpff,

At a tutorial, I was wondering if we could add:

![schema_train_test](https://user-images.githubusercontent.com/42702466/182960753-b6575ea0-ccb7-49d3-bcef-6d2f3ff519ad.svg)

It's a vertorized image, so possibly it will be bigger in the tutorial.

martinwimpff commented on examples/plot_unified_validation_scheme.py:
```diff
@@ -0,0 +1,344 @@
+# for this Hyperparamter search. Therefore we need a third split, the
+# so called validation set which is a Subset of the ``train_set``.
+# This second option splits the original ``train_set`` only once
+# (instead of k times as in Option 3) to speed up the tuning process.
```
We can do that but Robins method was a little different. He used Option 2 for tuning and Option 1 for the final evaluation (to use the full training set).

martinwimpff commented:
@sliwy: first of all, thanks a lot for your review!
> One more thing, now this tutorial and [`plot_hyperparameter_tuning_with_scikit-learn.py` ](https://github.com/braindecode/braindecode/blob/66ddda17875bb4d5b788228dd7a3b0b6117290b8/examples/plot_hyperparameter_tuning_with_scikit-learn.py) are quite similar. Maybe we can add `hyperparameters search` to the title of this one and just keep this tutorial? What do you think?

I agree with that, there shouldn't be two tutorials. The new title could be something like: How to train, test and tune your model. But I am open for your suggestions ;)

As you suggested I added additional parts for option 2 and 3 that do not use GridSearch.

@bruAristimunha: thanks for your suggestions as well, I added them as well.
Regarding the visualization, I also agree with both of you, we should include that. The image you posted looks really nice but in my opinion, something simpler would probably make the point clearer. Here is a quick example of what I had in mind:

![CV](https://user-images.githubusercontent.com/42936990/183027653-eaf125c6-f621-4473-b58d-aaf6e80d9344.png)

agramfort commented:
I would produce images with matplotlib like sklearn does to demo the cross validation generators

sliwy commented:
@martinwimpff thanks for the corrections!

`How to train, test and tune your model` - works for me :)

Having visualization done with matplotlib and sklearn could be also awesome.

Can we find some names for CV options (it is easier to use outside of tutorial context), my proposition:
1. Option 1 - train-test split
2. Option 2 - train-valid-test split
3. Option 3 - k-fold blockwise CV
In the case inside k-fold blockwise CV there is another one for hyperparams this could be indicated with adding 'nested'?

@martinwimpff For the option 3 visualization I would add `test_set` as a supplementary option because in some articles CV result is the final one while hyperparams are optimized in a nested CV. 

I took a while (quite short, so I just scanned paper for CV method) to prepare summary of evaluation methods used for models that are currently available in braindecode. Maybe it can be useful for discussion and showing how it is used in papers. Feel free to include it in the tutorial if you think it may be useful. I feel that it would be nice if we can express all the methods from this table in terms of our evaluation tutorial terminology. What do you think?

|          Article          |            Models             |    Hyperparameters optimization    |               Final evaluation                |
| :-----------------------: | :---------------------------: | :--------------------------------: | :-------------------------------------------: |
| Schirrmeister et al. 2017 | ShallowFBSCPNet<br />Deep4Net |              Option 2              |                   Option 1                    |
|    Lawhern et al. 2018    |            EEGNet             |                n/a                 | Option 3 (hold-out test only for SMR dataset) |
|    Chambon et al. 2018    |    SleepStagerChambon2018     |              Nested option 3              |        Option 3, no hold-out test set         |
|    Eldele et al. 2021     |     SleepStagerEldele2021     | Option 3 (without hold-out test)  |       Option 3 (without hold-out test)        |
|      Perslev et al.       |            USleep             |              Option 1              |           Option 1 on hold-out test           |



Schirrmeister, R.T., Springenberg, J.T., Fiederer, L.D.J., Glasstetter, M., Eggensperger, K., Tangermann, M., Hutter, F., Burgard, W. and Ball, T. (2017), Deep learning with convolutional neural networks for EEG decoding and visualization. Hum. Brain Mapp., 38: 5391-5420. https://doi.org/10.1002/hbm.23730

Lawhern, V.J., Solon, A.J., Waytowich, N.R., Gordon, S.M., Hung, C.P., & Lance, B. (2018). EEGNet: a compact convolutional neural network for EEG-based brain–computer interfaces. *Journal of Neural Engineering, 15*.

Chambon, S., Galtier, M.N., Arnal, P.J., Wainrib, G., & Gramfort, A. (2018). A Deep Learning Architecture for Temporal Sleep Stage Classification Using Multivariate and Multimodal Time Series. *IEEE Transactions on Neural Systems and Rehabilitation Engineering, 26*, 758-769.

Eldele, E., Chen, Z., Liu, C., Wu, M., Kwoh, C., Li, X., & Guan, C. (2021). An Attention-Based Deep Learning Approach for Sleep Stage Classification With Single-Channel EEG. *IEEE Transactions on Neural Systems and Rehabilitation Engineering, 29*, 809-818.

Perslev, M., Darkner, S., Kempfner, L., Nikolic, M., Jennum, P.J., & Igel, C. (2021). U-Sleep: resilient high-frequency sleep staging. *NPJ Digital Medicine, 4*.

martinwimpff commented:
> ``How to train, test and tune your model`` - works for me :)

👍 

> Can we find some names for CV options (it is easier to use outside of tutorial context), my proposition:
>
>1. Option 1 - train-test split
>2. Option 2 - train-valid-test split
>3. Option 3 - k-fold blockwise CV
In the case inside k-fold blockwise CV there is another one for hyperparams this could be indicated with adding 'nested'?

Agree. I am not sure about the nested CV. What we do in option 3 is a k-fold blockwise CV (with a fixed Hold-Out set, the untouched test set). As far as I understand this should be used with BCI research as the test set should always be from an unseen session or subject. Further, most datasets provide this two-fold split anyways. 
With nested CV one would throw everything in one data set and run two loops: one outer loop for splitting out the test set and one inner loop for the validation set. I don't really see a benefit from doing that for BCI research, as one would have to be extremely careful with how to design the splitters and the improvement is [quite marginal](https://scikit-learn.org/stable/auto_examples/model_selection/plot_nested_cross_validation_iris.html#sphx-glr-auto-examples-model-selection-plot-nested-cross-validation-iris-py).

> @martinwimpff For the option 3 visualization I would add test_set as a supplementary option because in some articles CV result is the final one while hyperparams are optimized in a nested CV.

I actually had the same thought while creating the image. Maybe we have to make the **division between HP search and "normal" training with 3 splits clearer**. For HP tuning the test set should never be used. 
For option 2 it may be used if you need to use the validation set for i.e. EarlyStopping. Else (if all HPs are fixed) option 1 should be used. 
The same holds for option 3 with the addition that the final test score is the average over k test scores.

We can include the table that you created in something like an additional section that we call `Reporting` or something like that? There we could also highlight the (large) differences between the publications + the occasional lack of documentation regarding the tuning/evaluation to further motivate why this page was created.

agramfort commented:
To avoid confusing the community I would aim to follow sklearn naming conventions and refer to sklearn documentation as much as possible. Maintaining a good documentation takes energy so let’s put our energy where we are very unique

bruAristimunha commented:
> I would produce images with matplotlib like sklearn does to demo the cross validation generators

Something like that @agramfort?

https://scikit-learn.org/stable/auto_examples/model_selection/plot_cv_indices.html

bruAristimunha commented:
> gree. I am not sure about the nested CV. What we do in option 3 is a k-fold blockwise CV (with a fixed Hold-Out set, the untouched test set). As far as I understand this should be used with BCI research as the test set should always be from an unseen session or subject. Further, most datasets provide this two-fold split anyways. With nested CV one would throw everything in one data set and run two loops: one outer loop for splitting out the test set and one inner loop for the validation set. I don't really see a benefit from doing that for BCI research, as one would have to be extremely careful with how to design the splitters and the improvement is [quite marginal](https://scikit-learn.org/stable/auto_examples/model_selection/plot_nested_cross_validation_iris.html#sphx-glr-auto-examples-model-selection-plot-nested-cross-validation-iris-py).
> 
> > @martinwimpff For the option 3 visualization I would add test_set as a supplementary option because in some articles CV result is the final one while hyperparams are optimized in a nested CV.
> 
> I actually had the same thought while creating the image. Maybe we have to make the **division between HP search and "normal" training with 3 splits clearer**. For HP tuning the test set should never be used. For option 2 it may be used if you need to use the validation set for i.e. EarlyStopping. Else (if all HPs are fixed) option 1 should be used. The same holds for option 3 with the addition that the final test score is the average over k test scores.
> 
> We can include the table that you created in something like an additional section that we call `Reporting` or something like that? There we could also highlight the (large) differences between the publications + the occasional lack of documentation regarding the tuning/evaluation to further motivate why this page was created.

bruAristimunha commented:
> > `How to train, test and tune your model` - works for me :)
> 
+2. I love the new title =)
> 
> > Can we find some names for CV options (it is easier to use outside of tutorial context), my proposition:
> > 
> > 1. Option 1 - train-test split
> > 2. Option 2 - train-valid-test split
> > 3. Option 3 - k-fold blockwise CV
> >    In the case inside k-fold blockwise CV there is another one for hyperparams this could be indicated with adding 'nested'?
> 
> Agree. I am not sure about the nested CV. What we do in option 3 is a k-fold blockwise CV (with a fixed Hold-Out set, the untouched test set). As far as I understand this should be used with BCI research as the test set should always be from an unseen session or subject. Further, most datasets provide this two-fold split anyways. With nested CV one would throw everything in one data set and run two loops: one outer loop for splitting out the test set and one inner loop for the validation set. I don't really see a benefit from doing that for BCI research, as one would have to be extremely careful with how to design the splitters and the improvement is [quite marginal](https://scikit-learn.org/stable/auto_examples/model_selection/plot_nested_cross_validation_iris.html#sphx-glr-auto-examples-model-selection-plot-nested-cross-validation-iris-py).

> > @martinwimpff For the option 3 visualization I would add test_set as a supplementary option because in some articles CV result is the final one while hyperparams are optimized in a nested CV.
> 
> I actually had the same thought while creating the image. Maybe we have to make the **division between HP search and "normal" training with 3 splits clearer**. For HP tuning the test set should never be used. For option 2 it may be used if you need to use the validation set for i.e. EarlyStopping. Else (if all HPs are fixed) option 1 should be used. The same holds for option 3 with the addition that the final test score is the average over k test scores.
> 
> We can include the table that you created in something like an additional section that we call `Reporting` or something like that? There we could also highlight the (large) differences between the publications + the occasional lack of documentation regarding the tuning/evaluation to further motivate why this page was created.

I love the idea of a reporting subsection to discuss a little the potential difference that we can find in metrics reported in articles and during the replication moment. I think you just passed this with deep4net, right @martinwimpff?

I would love this table in my thesis, but not in the library. With more deep learning methods, it may become unfeasible to maintain the table.

I wonder if we're putting too much information into a tutorial. I would suggest that we don't spend too much energy on definitions, as indicated by @agramfort, and close with a simple matplotlib view, like https://scikit-learn.org/stable/auto_examples/model_selection/plot_cv_indices.html. What do you think, @martinwimpff, @sliwy and @agramfort?

martinwimpff commented:
> To avoid confusing the community I would aim to follow sklearn naming conventions and refer to sklearn documentation as much as possible. Maintaining a good documentation takes energy so let’s put our energy where we are very unique

I think that this is a valid point and it might be better if we reorganize the content of this tutorial, also regarding the merge with the current hyperparameter tuning tutorial.

So, as mentioned above, I would like to make the division between "normal" training and HP tuning clearer. Therefore I would propose the following struture:

1. Introduction (motivate the topic)
2. How to train and evaluate your model
2.1 Simple Train-Test Split
2.2 Train-Val-Test Split
2.3 (blockwise) k-Fold CV (with Holdout test set)
3. Hyperparameter Tuning
3.1 Train-Val(-Test) Split
3.2 (blockwise) k-Fold CV
4. Reporting

As @agramfort mentioned, we should reference the sklearn library as often as possible (especially in sections 2 and 3). Further, the explanations of the schemes should be minimal and rather focus on the specifics of BCI data (e.g. test set should be from another session or subject, train-val split should be blockwise, is there anything else?). Is this what you meant with too much information @bruAristimunha ?

bruAristimunha commented:
Yes @martinwimpff

sliwy commented:
@bruAristimunha I agree, the table may be too much for the tutorial. The matplotlib simple plots look good to me! :)

@martinwimpff I like the new structure of the tutorial. Just for point 2.3 I would mention that it can be used as well without hold-out test set.

martinwimpff commented:
I just committed a new version with the restructuring mentioned above.
@sliwy @bruAristimunha what do you think of this?
@bruAristimunha How should we proceed with the visualizations? Have you already had a look into it?

bruAristimunha commented:
Hello @martinwimpff and @sliwy,
I have included three visualizations. Can you look and review it?

martinwimpff commented:
Great job @bruAristimunha!

I have two small suggestions for the visualizations:
1. I would keep the colors consistent throughout the images (currently the test set color changes from 1 to 2&3)
2. I would keep the scaling for option 3 (x-axis ratio) s.t. its clearer that the test set (size) remains the same in all 3 options. You probably wanted to make the kFold visualization as big as possible + clearly separate the test set. So this one is probably a matter of taste ;)

sliwy commented on examples/plot_how_train_test_and_tune.py:
```diff
@@ -0,0 +1,669 @@
+        yticks=[0.5, 3.5],
+        yticklabels=["Train-Test\nSplit", "Original\nDataset"],
+        xlabel="Number of samples.",
+        title="Training-Testing Split",
```
```suggestion
        title="Train-Test Split",
```

sliwy reviewed (approved):
Great job @martinwimpff @bruAristimunha! Thanks for that!

The only small thing is that the code for generating the visualization complicates the tutorial a little bit, but I don't have any idea how to fix that apart from moving those functions to some module. I am not a big fan of this option either.

To me, after the small changes in the visualization mentioned by @martinwimpff it is ready to be merged! :)

bruAristimunha commented:
I have incorporated all comments. For me, we can apply merge, @agramfort.

sliwy commented:
@bruAristimunha I almost forgot, do we finally remove the [plot_hyperparameter_tuning_with_scikit-learn.py](https://github.com/braindecode/braindecode/blob/66ddda17875bb4d5b788228dd7a3b0b6117290b8/examples/plot_hyperparameter_tuning_with_scikit-learn.py)?

bruAristimunha commented:
Can we handle this in another pull request?

agramfort commented on examples/plot_how_train_test_and_tune.py:
```diff
@@ -0,0 +1,669 @@
+will not further elaborate on these parts and you can feel free to skip them.
+
+In general we distinguish between "usual" training and evaluation and hyperparameter search.
+The Tutorial is therefore split into two parts, one for the three different training schemes
```
```suggestion
The tutorial is therefore split into two parts, one for the three different training schemes
```

agramfort commented on examples/plot_how_train_test_and_tune.py:
```diff
@@ -0,0 +1,669 @@
+######################################################################
+# Why should I care about model evaluation?
+# -----------------------------------------
+# Short answer: To produce reliable results!
```
```suggestion
# Short answer: Yes to produce reliable results!
```

agramfort commented on examples/plot_how_train_test_and_tune.py:
```diff
@@ -0,0 +1,669 @@
+# data into two parts, training and testing sets. It sounds like a
+# simple division, right? But the story does not end here.
+#
+# While developing a ML model you
+# usually have to adjust and tune hyperparameters of your model or
+# pipeline (e.g., number of layers, learning rate, number of epochs).
+# Deep learning models usually have many free parameters; they could
+# be considered complex models with many degrees of freedom.
+# If you kept using the test dataset to evaluate your adjustment
+# you would run into data leakage.
```
can you format this to leverage the full authorized line width (80c ?)

agramfort commented on examples/plot_how_train_test_and_tune.py:
```diff
@@ -0,0 +1,669 @@
+# This means that if you use the test set to adjust the hyperparameters
+# of your model, the model implicitly learns or memorizes the test set.
+# Therefore, the trained model is no longer independent of the test set
+# (even though they were never used for backpropagation!).
```
```suggestion
# (even though they were never used for training!).
```

agramfort commented on examples/plot_how_train_test_and_tune.py:
```diff
@@ -0,0 +1,669 @@
+# of your model, the model implicitly learns or memorizes the test set.
+# Therefore, the trained model is no longer independent of the test set
+# (even though they were never used for backpropagation!).
+# If you perform any Hyperparameter tuning, you need a third split,
```
```suggestion
# If you perform any hyperparameter tuning, you need a third split,
```

agramfort commented on examples/plot_how_train_test_and_tune.py:
```diff
@@ -0,0 +1,669 @@
+
+n_classes = 4
+# Extract number of chans and time steps from dataset
+n_chans = windows_dataset[0][0].shape[0]
```
for later but n_channels would be nicer for me

agramfort commented on examples/plot_how_train_test_and_tune.py:
```diff
@@ -0,0 +1,669 @@
+# ``Subset`` from torch and ``predefined_split`` from skorch.
+#
+
+from torch.utils.data import Subset
+from sklearn.model_selection import train_test_split
+from skorch.helper import predefined_split, SliceDataset
+
+X_train = SliceDataset(train_set, idx=0)
+y_train = np.array([y for y in SliceDataset(train_set, idx=1)])
+train_indices, val_indices = train_test_split(
+    X_train.indices_, test_size=0.2, shuffle=False
+)
```
this is neat but a bit convoluted for an intro tutorial.

I would consider exposing in train_set indices or something that can simply
be passed to train_test_split

martinwimpff replied:
One could use
```
train_test_split(np.array(len(train_set)), test_size=0.2, shuffle=False)
```
instead, but `X_train` and `y_train` would still be necessary for `cross_val_score, search.fit()` and the visualizations.

sliwy commented:
@bruAristimunha sure!

Once again, thanks for this great tutorial! 👏

agramfort commented on examples/plot_how_train_test_and_tune.py:
```diff
@@ -0,0 +1,669 @@
+# `cross_val_score <https://scikit-learn.org/stable/modules/generated/
+# sklearn.model_selection.cross_val_score.html>`__ and the `KFold
+# <https://scikit-learn.org/stable/modules/generated/sklearn.model_
+# selection.KFold.html>`__. cross-validator.
```
its not called a cross-validator in sklearn but a cross-validation splitter

martinwimpff replied:
I had the same thought about that one as they call it CV generator or CV splitter in general but cross-validator
In the [KFold doc](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.KFold.html).
The class itself also inherits from `BaseCrossValidator`. 
How should we call it?

agramfort commented on examples/plot_how_train_test_and_tune.py:
```diff
@@ -0,0 +1,669 @@
+    test_set=test_set,
+)
+######################################################################
+# How to tune your Hyperparameters
```
```suggestion
# How to tune your hyperparameters
```

agramfort commented on examples/plot_how_train_test_and_tune.py:
```diff
@@ -0,0 +1,669 @@
+train_val_split = KFold(n_splits=5, shuffle=False)
+
+######################################################################
+# .. note::
```
this is something that should appear at the top and maybe as a warning as it's not a detail.

sliwy replied:
maybe we can use `copy.deepcopy(model)` in every `EEGClassifier` init, so we won't modify the weights of original model?

martinwimpff replied:
I prefer the warning option. If we use `deepcopy` we would have to include a small note on that as well to avoid confusion.

agramfort commented:
arff I would use the word CV splitter

> Message ID: ***@***.***
> com>
>

martinwimpff commented:
I included the comments from @agramfort.

I did not change this part:
```python
from torch.utils.data import Subset
from sklearn.model_selection import train_test_split
from skorch.helper import predefined_split, SliceDataset

X_train = SliceDataset(train_set, idx=0)
y_train = np.array([y for y in SliceDataset(train_set, idx=1)])
train_indices, val_indices = train_test_split(
    X_train.indices_, test_size=0.2, shuffle=False
)
```
as `X_train` and `y_train` are used multiple times later in the tutorial. 

I changed the final note to a warning and moved it to the start as I personally prefer this option for the tutorial. I also like the `deepcopy` option in general (since it it a very common thing to do) but I think that it might confuse people in this tutorial.

agramfort reviewed (approved):
looks good !

https://output.circle-artifacts.com/output/job/218ce5d3-5b86-4094-a596-ddce84388a1a/artifacts/0/dev/auto_examples/plot_how_train_test_and_tune.html

agramfort commented:
thanks @martinwimpff @bruAristimunha 🙏

bruAristimunha commented:
A huge thank you to you, @martinwimpff. For building the entire tutorial! Welcome to the braindecode contributor group, and we look forward to your next contributions! Always feel welcome to contribute and interact with the library.

robintibor commented:
Also wanted to say a big thank you to @martinwimpff  @bruAristimunha and also @agramfort the tutorial really looks incredible, really a valuable addition to the library. I had been a bit absent and feel super happy to see such great work once I found the time to look haha :)
